### PR TITLE
[Security][SecurityBundle] Add voter individual decisions to profiler

### PR DESCRIPTION
**Pull request by l-vo**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no 
| Tests pass?   | yes
| Fixed tickets |
| License       | MIT
| Doc PR        | 

On the profiler (security tab), the decisions made by the AccessDecisionManager are displayed with the voters registered in the application. But there is no information about which voter was really used for each AccessDecisionManager choice and which result the voters returned.

This PR allows to display for each AccessDecisionManager choice, the voters implicated and the vote they returned.

Screenshot profiler before PR:
capture d ecran 2018-07-10 a 13 26 46
Screenshot profiler after PR:
capture d ecran 2018-10-10 a 21 35 30

### CODE REVIEW

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/Compiler/AddSecurityVotersPass.php
Comment (l-vo) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/Compiler/AddSecurityVotersPass.php]: I'm not sure it's the best approach. Is it better to test `kernel.debug` here, or create a specific compiler pass only added if `kernel.debug` is set to true ?

**Comment by javiereguiluz** [comment]
@l-vo this is something we definitely want to improve. Thanks!  If possible, please paste a before/after screenshot comparison.

**Comment by l-vo** [comment]
I didn't understand what CHANGELOG file I should update ? There is not CHANGELOG.md file and the CHANGELOG-4.1.md not seems to me the good file since I think my PR is for the 4.2 version.

File: src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php
Comment (curry684) [inline src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php]: * `$reflectionClass`

**Comment by l-vo** [comment]
@javiereguiluz : screenshots added, thanks !

File: src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php
Comment (xabbuh) [inline src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php]: I am not convinced that we really need to use reflection here. Doing some hardcoded checks below is probably good as well.

File: src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php
Comment (xabbuh) [inline src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php]: Removing the method is a BC break I'd try to avoid.

File: src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php
Comment (l-vo) [reply src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php]: @curry684 Fixed, thank you :)

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/Compiler/AddSecurityVotersPass.php
Comment (chalasr) [reply src/Symfony/Bundle/SecurityBundle/DependencyInjection/Compiler/AddSecurityVotersPass.php]: ok as is to me

Comment (l-vo) [reply src/Symfony/Bundle/SecurityBundle/DependencyInjection/Compiler/AddSecurityVotersPass.php]: @chalasr ok, thanks for your help

File: src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php
Comment (l-vo) [reply src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php]: @xabbuh Modified with hardcoded values, thanks

File: src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php
Comment (l-vo) [reply src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php]: @xabbuh I Restored the method with a deprecated flag

File: src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php
Comment (ro0NL) [inline src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php]: do we use it still? Should we trigger a hard deprecation using trigger_error()?

Comment (l-vo) [reply src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php]: Right ! I added it, thanks :)

**Comment by ro0NL** [comment]
I'd like to see the "Show/hide voter details" toggle per row in the initial table (right column), toggling a single new table of voter details underneath it. Yes, there'd be always 1 active table with voter details.

Currently it looks cluttered, and unclear which table belongs to what.

**Comment by javiereguiluz** [comment]
@ro0NL that's exactly what I was thinking. It's like the "Show more" link that GitHub uses in some places:

show-more

My idea was going to wait until this PR was merged and then tweak its design a bit in another PR. I have some sketches ready, but need time to polish it.

**Comment by l-vo** [comment]
Add PR changes to CHANGELOG files

**Comment by l-vo** [comment]
@ro0NL @javiereguiluz ok for doing that in this way, I was not really inspired for how changing the current interface efficiently.

**Comment by l-vo** [comment]
Branch rebased and conflicts resolved.

**Comment by fabpot** [comment]
@ro0NL @javiereguiluz Can you chime in here about the UI?

File: src/Symfony/Bundle/SecurityBundle/Resources/views/Collector/security.html.twig
Comment (rpkamp) [inline src/Symfony/Bundle/SecurityBundle/Resources/views/Collector/security.html.twig]: Checking against the constants here would be more explicit, i.e.,

```
voter_detail['vote'] == constant('Symfony\\Component\\Security\\Core\\Authorization\\Voter\\VoterInterface::ACCESS_GRANTED')
```

File: src/Symfony/Bundle/SecurityBundle/Resources/views/Collector/security.html.twig
Comment (rpkamp) [inline src/Symfony/Bundle/SecurityBundle/Resources/views/Collector/security.html.twig]: Why not put the value here for easy debugging if this ever happens? `unknown ({{ voter_detail['vote'] }})`

**Review by rpkamp — COMMENTED** [review]
Looks awesome. Just found 2 very minor points.

File: src/Symfony/Bundle/SecurityBundle/Resources/views/Collector/security.html.twig
Comment (l-vo) [reply src/Symfony/Bundle/SecurityBundle/Resources/views/Collector/security.html.twig]: @rpkamp  It make sense, modified, thanks !

File: src/Symfony/Bundle/SecurityBundle/Resources/views/Collector/security.html.twig
Comment (l-vo) [reply src/Symfony/Bundle/SecurityBundle/Resources/views/Collector/security.html.twig]: @rpkamp Yes ! Updated thanks :)

**Comment by l-vo** [comment]
@ro0NL @javiereguiluz I slightly improved the UI.

**Comment by l-vo** [comment]
Deprecation flag is not needed anymore.

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/Compiler/AddSecurityVotersPass.php
Comment (xabbuh) [reply src/Symfony/Bundle/SecurityBundle/DependencyInjection/Compiler/AddSecurityVotersPass.php]: but we IMO don't need the `hasParameter()` call here

File: src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php
Comment (xabbuh) [reply src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php]: I think we should rather save this function call at all and iterate `$this->accessDecisionManager->getVoters()` instead to populate `$this->data['voters']` (just like it was done before)

File: src/Symfony/Component/Security/CHANGELOG.md
Comment (l-vo) [reply src/Symfony/Component/Security/CHANGELOG.md]: Thanks :)

File: src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php
Comment (l-vo) [reply src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php]: Modification done :)

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/Compiler/AddSecurityVotersPass.php
Comment (l-vo) [reply src/Symfony/Bundle/SecurityBundle/DependencyInjection/Compiler/AddSecurityVotersPass.php]: I did it in this way because it was like that in `SecurityExtension.php`. So I modified it in `SecurityExtension.php` and `AddSecurityVoterPass.php` and fixed the tests that change broke.

**Comment by l-vo** [comment]
@xabbuh I fixed the things you reported. Could you (or anyone else) remove the deprecated flag which is not needed anymore ? Thanks in advance :)

File: src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php
Comment (xabbuh) [inline src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php]: This is always an array, isn't it?

File: src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php
Comment (xabbuh) [inline src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php]: can be removed

File: src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php
Comment (xabbuh) [inline src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php]: We *could* have kept it here and just use `$this->data['access_decision_log']` below to keep the diff a bit more compact.

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/Compiler/AddSecurityVotersPass.php
Comment (xabbuh) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/Compiler/AddSecurityVotersPass.php]: I wonder if those ids could potentially conflict with services from the DebugBundle and if we should rather use something like `security.voter.debug.` as the prefix instead.

File: src/Symfony/Bundle/SecurityBundle/Tests/DataCollector/SecurityDataCollectorTest.php
Comment (xabbuh) [inline src/Symfony/Bundle/SecurityBundle/Tests/DataCollector/SecurityDataCollectorTest.php]: maybe use `AccessDecisionManagerInterface` instead here

File: src/Symfony/Bundle/SecurityBundle/Tests/DataCollector/SecurityDataCollectorTest.php
Comment (xabbuh) [inline src/Symfony/Bundle/SecurityBundle/Tests/DataCollector/SecurityDataCollectorTest.php]: `assertSame()`

File: src/Symfony/Bundle/SecurityBundle/Tests/DataCollector/SecurityDataCollectorTest.php
Comment (xabbuh) [inline src/Symfony/Bundle/SecurityBundle/Tests/DataCollector/SecurityDataCollectorTest.php]: I think we should better use `assertCount()` with an expectation of `0`

File: src/Symfony/Bundle/SecurityBundle/Tests/DataCollector/SecurityDataCollectorTest.php
Comment (xabbuh) [inline src/Symfony/Bundle/SecurityBundle/Tests/DataCollector/SecurityDataCollectorTest.php]: same here

**Review by xabbuh — COMMENTED** [review]
This looks really good to me. I think we are almost there. I just left few remaining comments. Great work so far! 👍

File: src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php
Comment (l-vo) [reply src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php]: I don't think so, original data seems to be transformed into a Data object when lateCollect is invoked. Are you agree with that ?

File: src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php
Comment (l-vo) [reply src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php]: Ok, done

File: src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php
Comment (l-vo) [reply src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php]: No problem, done.

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/Compiler/AddSecurityVotersPass.php
Comment (l-vo) [reply src/Symfony/Bundle/SecurityBundle/DependencyInjection/Compiler/AddSecurityVotersPass.php]: It makes sense, done :)

File: src/Symfony/Bundle/SecurityBundle/Tests/DataCollector/SecurityDataCollectorTest.php
Comment (l-vo) [reply src/Symfony/Bundle/SecurityBundle/Tests/DataCollector/SecurityDataCollectorTest.php]: Indeed, thanks !

**Comment by l-vo** [comment]
@xabbuh thanks :), last reports fixed except the return of `getVoterDetails` but I'm not sure my implementation is wrong.

File: src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php
Comment (l-vo) [reply src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php]: @xabbuh I set this discussion as unresolved since `$this->accessDecisionManager->getVoters()` returns `TraceableVoter` instances, so `TraceableVoter` is displayed in the profiler instead of `ExpressionVoter` for instance. It was the reason I did that in this way. I'm going to fix it :)

Comment (l-vo) [reply src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php]: Fixed and tests added.

**Review by xabbuh — DISMISSED** [review]
needs some conflicts to get resolved though :)

**Comment by l-vo** [comment]
@xabbuh conflicts resolved :)

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/Compiler/AddSecurityVotersPass.php
Comment (chalasr) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/Compiler/AddSecurityVotersPass.php]: Not needed, it’s private by default and decorators should have the same visibility level as the decorated

Comment (stof) [reply src/Symfony/Bundle/SecurityBundle/DependencyInjection/Compiler/AddSecurityVotersPass.php]: well, a decorator service should be private, as its own id will never be used for dynamic access (even when decorating a public service, the decorator service is still private. DecoratorServicePass takes care that the decorated service id keeps the same visibility).

But it is true that being private is the default for a definition.

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/Compiler/AddSecurityVotersPass.php
Comment (stof) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/Compiler/AddSecurityVotersPass.php]: this array is useless. It is exactly the same than `$voters` in the end

File: src/Symfony/Component/Security/Core/Authorization/TraceableAccessDecisionManager.php
Comment (stof) [inline src/Symfony/Component/Security/Core/Authorization/TraceableAccessDecisionManager.php]: this breaks the lazy-loading of voters, as it will now load all voters even though the original code was not triggering their loading

File: src/Symfony/Component/Security/Core/Authorization/Voter/TraceableVoter.php
Comment (stof) [inline src/Symfony/Component/Security/Core/Authorization/Voter/TraceableVoter.php]: that's not really a log, as there is a single value

File: src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php
Comment (xabbuh) [reply src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php]: I don't think that's needed. Thanks for checking it btw. :+1:

File: src/Symfony/Component/Security/Core/Authorization/TraceableAccessDecisionManager.php
Comment (stof) [inline src/Symfony/Component/Security/Core/Authorization/TraceableAccessDecisionManager.php]: This implementation won't work fine in many cases:

- if a voter is not called at all for this decision (only the `consensus` strategy is guaranteed to always call all voters. `affirmative` has a shortcut when granting access, and `unanimous` has one when denying access), you will retrieve the log of a previous decision
- if a voter triggers another decision internally, the log being retrieved here will be the wrong one

**Review by stof — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Security/Core/Authorization/TraceableAccessDecisionManager.php
Comment (l-vo) [reply src/Symfony/Component/Security/Core/Authorization/TraceableAccessDecisionManager.php]: Arf, I didn't think about that at all... Thanks !

File: src/Symfony/Component/Security/Core/Authorization/TraceableAccessDecisionManager.php
Comment (l-vo) [reply src/Symfony/Component/Security/Core/Authorization/TraceableAccessDecisionManager.php]: @stof thank you for your review, I'm going to think again about my implementation. However one of your comments is not very clear for me:
> **if a voter triggers another decision internally**, the log being retrieved here will be the wrong one

How is it possible ?

**Comment by l-vo** [comment]
@stof implementation reworked, thanks !

**Comment by l-vo** [comment]
Status: Needs Review

**Comment by l-vo** [comment]
Status: Needs Work

**Comment by l-vo** [comment]
Status changed back to needs work, missing the edge case of unanimous strategy where the voter is called for each attribute. (Need to display voter attribute in the details on this case).

**Comment by l-vo** [comment]
Screenshot with unanimous voter:
capture d ecran 2018-10-11 a 13 33 12

**Comment by l-vo** [comment]
Status: Needs review

File: src/Symfony/Component/Security/Core/Authorization/AccessDecisionManager.php
Comment (fabpot) [inline src/Symfony/Component/Security/Core/Authorization/AccessDecisionManager.php]: Votes and dispatches event.

File: src/Symfony/Bundle/SecurityBundle/Tests/EventListener/VoteListenerTest.php
Comment (fabpot) [inline src/Symfony/Bundle/SecurityBundle/Tests/EventListener/VoteListenerTest.php]: I would remove the phpdoc here as well

File: src/Symfony/Bundle/SecurityBundle/Tests/EventListener/VoteListenerTest.php
Comment (fabpot) [inline src/Symfony/Bundle/SecurityBundle/Tests/EventListener/VoteListenerTest.php]: I would remove the docblock, we don't a description of the test class

**Review by fabpot — CHANGES_REQUESTED** [review]
I'm wondering if dispatching an event for each vote has a significant perf impact (as some apps might have a large number of votes for a page).

File: src/Symfony/Component/Security/Core/Authorization/TraceableAccessDecisionManager.php
Comment (stof) [reply src/Symfony/Component/Security/Core/Authorization/TraceableAccessDecisionManager.php]: Well, you can inject the AccessDecisionManager inside your voter (we made that possible in 3.3 or 3.4, I dont remember)

Comment (l-vo) [reply src/Symfony/Component/Security/Core/Authorization/TraceableAccessDecisionManager.php]: Hum indeed... So in this case, the profiler displays the decision of the ADM  called by the voter with the other decisions. This behavior seems unexpected to me. Do you think it's a correct behavior or does it deserve to be fixed ?

Comment (l-vo) [reply src/Symfony/Component/Security/Core/Authorization/TraceableAccessDecisionManager.php]: I created an other PR for that: #28855

**Comment by l-vo** [comment]
@fabpot PR reworked, the event is dispatched by a voter decorator now. This decorator is only created in debug mode.

**Comment by l-vo** [comment]
status:  needs review

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php
Comment (fabpot) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php]: should reverted

File: src/Symfony/Component/Security/Core/Authorization/TraceableAccessDecisionManager.php
Comment (fabpot) [inline src/Symfony/Component/Security/Core/Authorization/TraceableAccessDecisionManager.php]: Adds

File: src/Symfony/Component/Security/Core/VoteEvents.php
Comment (fabpot) [inline src/Symfony/Component/Security/Core/VoteEvents.php]: I would remove this class and "harcode" the event name as we don't want people to listen to this event (which is purely internal and only available in debug mode anyway).

**Review by fabpot — CHANGES_REQUESTED** [review]
Looks much better now. Made some small suggestions.

**Review by ro0NL — COMMENTED** [review]
should `VoteEvent` + `VoteListener` also marked `@internal`? Tend to believe at least the event should kept internal.

**Comment by l-vo** [comment]
@ro0NL It makes sense, done, thanks :)
@fabpot Last modifications done, thanks !

**Comment by l-vo** [comment]
status: needs review

**Comment by l-vo** [comment]
@chalasr IMO It is fully resolved. This problem occured when `TraceableVoter` stored consecutive votes of its decorated voter. Using events and storing this information at AccessDecisionManager level solves this issue.

**Review by chalasr — APPROVED** [review]
Good for 4.2 IMHO

File: src/Symfony/Bundle/SecurityBundle/EventListener/VoteListener.php
Comment (l-vo) [inline src/Symfony/Bundle/SecurityBundle/EventListener/VoteListener.php]: from traceable voters

File: src/Symfony/Component/Security/CHANGELOG.md
Comment (l-vo) [inline src/Symfony/Component/Security/CHANGELOG.md]: (only in debug mode)

Comment (ro0NL) [reply src/Symfony/Component/Security/CHANGELOG.md]: not sure it should even be mentioned as it's internal

Comment (l-vo) [reply src/Symfony/Component/Security/CHANGELOG.md]: Right ! Thank you :)

**Comment by l-vo** [comment]
Fixed some minors things, `VoteListener` class description and Security `CHANGELOG` (see self-review above). No code modified.

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @l-vo.

**Pull request merged**
